### PR TITLE
KAFKA-5154: Consumer fetches from revoked partitions when SyncGroup fails with disconnection [WIP]

### PR DESCRIPTION
**Pull request by dguy**
Scenario is as follows:
1. Consumer subscribes to topic t1 and begins consuming
2. heartbeat fails as the group is rebalancing
3. ConsumerCoordinator.onJoinGroupPrepare is called
   3.1 onPartitionsRevoked is called
4. consumer becomes the group leader
5. sends sync group request
6. sync group is cancelled due to disconnection
7. fetch request is sent for partitions that have previously been revoked

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @hachikuji this is not to be merged, but this test fails and is based on the logs i extracted from the corresponding JIRA.
One simple fix for this specific problem is to clear the `subscription` in `ConsumerCoordinator.onJoinPrepare`. I think that is probably worth doing, but it is possibly masking a bigger problem as once the SyncGroup is disconnected the rebalance is never completed, i.e., the consumer doesn't get any partitions assigned

**Comment by guozhangwang** [comment]
Regarding the general issue that it may get exposed: I think the root cause is in `CoordinatorResponseHandler#onFailure()` we only mark the coordinator as dead but do not do anything else; for example for `syncGroupRequest`, and then in its caller `joinGroupIfNeeded()`, we do not check the `DisconnectException` again, and the while condition `needRejoin() || rejoinIncomplete()` will also pass although the actual state is:

1. `state = MemberState.UNJOINED`.
2. `coordinator = null`.
3. `rejoinNeeded = false`.
4. `joinFuture = null`.

I think we should check all four variables above in the while condition instead.

**Comment by hachikuji** [comment]
Good catch. I wonder if we are setting `rejoinNeeded` to false too early. Perhaps that should only come after the `SyncGroup` returns?

**Comment by dguy** [comment]
@guozhangwang @hachikuji i **think** i've fixed it by simply calling `requestRejoin` when the coordinator is marked as dead. It will now attempt to rejoin the group. The test passes and everything looks as if it is working as expected

**Comment by guozhangwang** [comment]
Seems one fix missing:

```
scala2.11/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1271: error: no suitable constructor found for Metadata(int,long)
        Metadata metadata = new Metadata(0, Long.MAX_VALUE);
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -783,8 +783,10 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
         @Override
         public void onFailure(RuntimeException e, RequestFuture<T> future) {
             // mark the coordinator as dead
-            if (e instanceof DisconnectException)
+            if (e instanceof DisconnectException) {
                 coordinatorDead();
+                requestRejoin();
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I'm not sure if we should set `requestRejoin` in the base class (`CoordinatorResponseHandler `). For example, `HeartbeatResponseHandler` also extends from it, but for that request if we get a disconnect, we should just mark the coordinator as dead in order to re-discover it; and then after new coordinator rediscovered retry sending heartbeat request and if that succeed just proceed as normal. Setting it here will force heartbeat request disconnection to also trigger a join group.

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I think @hachikuji 's suggestion may be better: do not call 

```
AbstractCoordinator.this.rejoinNeeded = false;
```

in `JoinGroupResponseHandler#handle()`, but in `SyncGroupResponseHandler#handle()`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 2]
```diff
@@ -537,6 +536,7 @@ public void handle(SyncGroupResponse syncResponse,
             if (error == Errors.NONE) {
                 sensors.syncLatency.record(response.requestLatencyMs());
                 future.complete(syncResponse.memberAssignment());
+                AbstractCoordinator.this.rejoinNeeded = false;
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:539]: On a second thought... `future.complete(syncResponse.memberAssignment())` above will trigger `joinFuture.addListener`'s `onSuccess`, which will enable the heartbeat thread right away, and hence there is a (very small) race condition.

I think it is safer to just move the the above line inside `onSuccess` (line 395) to set it before enabling heart beat thread, and we would not need `AbstractCoordinator.this` prefix also.

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:539]: @dguy @hachikuji if it sounds good to you I can go ahead and make this change while merging.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:539]: LGTM

Comment (dguy) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:539]: LGTM

**Comment by guozhangwang** [comment]
Merged to trunk and cherry-picked to 0.11.0.

**Pull request closed without merging**